### ISSUE TITLE
multiple modals not working

### ISSUE THREAD

**Issue by macrozone** [react-native-community/react-native-modal#171]
I switched from react-native's default Modal to react-native-modal and i noticed that if I have multiple modals open at the same time, only one will ever be visible.

react-native's modal handles this correctly: they display only the first one, but if i dismiss, i see the other modals.

With react-native-modal, the others are simply ignored, altough they are mounted with `isVisible={true}`. This leads to inconsistency.

**Comment by mmazzarolo** [comment]
Hi! 
Unfortunately it is no possible showing more than a modal *at the same time* right now because react native's default modal do not support this feature.

The should be a few issues opened both here and in the react-native repo about this issue. It happens even with dialogs/alerts.

If, instead of showing the modals at the same time, you just want to show a modal one after another you should keep in mind that you should wait for the animation to complete before showing it (both with this modal with the original one).  

TLDR:
Show the new second modal using the `onModalHide` prop of the first one.

Let me know if you need further info!

**Comment by macrozone** [comment]
hey @mmazzarolo thanks for your answer.

i think you misunderstand my issue, so let me explain:

- react-native's default modal does not allow to show multiple modals at the same time. 
- HOWEVER, if you mount multiple modals, the first one will be visible and the others will become visible, once the first one has been dismissed. So it's guaranteed that the user will see all modals, altough not at the same time, but sequentially
- in react-native-modal, this is not the case: The others will simple be ignored, once the first one will be dismissed. This is a problem, because you will end up having multiple Modals mounted with visible=true that are actually not visible. This is inconsistent.
- `onModalHide` is not an option. Imaging having multiple containers that might show a modal, depending on some state. They don't know about each other. You would need to have a global modal-managment, and only have one root Modal. This is possible, but for simpler apps just too much hazzle.

**Comment by mmazzarolo** [comment]
Re-opening, sorry if I misunderstood it.


> HOWEVER, if you mount multiple modals, the first one will be visible and the others will become visible, once the first one has been dismissed. So it's guaranteed that the user will see all modals, altough not at the same time, but sequentially

Oh! I'm was not aware of this -- and I don't know if I understood it correctly, could you show me a snippet? Are those modals animated?
react-native-modal is using the original modal under the hood so I don't see why they should behave differently here 🤔 

> onModalHide is not an option. Imaging having multiple containers that might show a modal, depending on some state. They don't know about each other. You would need to have a global modal-managment, and only have one root Modal. This is possible, but for simpler apps just too much hazzle.

Well, this is actually is a known issue of react-native about modals, alerts and dialogs. 
There are multiple issues about this exact behaviour on both iOS and Android. 
[Here is the first one that comes to my mind](https://github.com/facebook/react-native/issues/10471).

**Comment by mmazzarolo** [comment]
Closing, feel free to write here again if you want to discuss it further and I'll re-open it 👍

**Comment by kalwaniyash** [comment]
@macrozone Hey, i was facing a similar issue and i wrote a wrapper on top of the react-native-modal that does exactly what you said. If one modal is already visible and another one needs to show itself, then the wrapper will close the first modal and show the second one automatically. The UI won't freeze and stop working.
Let me know if you still need it.

**Comment by macrozone** [comment]
hi guys,

I switched back to the native modal for the time beeing, but i might revisit this one here in the future. 

@kalwaniyash if you have a link to a repo or gist, do you mind sharing it? thank you!

**Comment by kalwaniyash** [comment]
@macrozone the code currently is deeply integrated with my project i will soon create a public repo and share it over here.

**Comment by mmazzarolo** [comment]
Re-opening 👍

**Comment by jojonarte** [comment]
@kalwaniyash what approach did you use.

**Comment by IlifilzaRusli** [comment]
@kalwaniyash I'm interested as well, can you tell us more?

**Comment by kalwaniyash** [comment]
@jojonarte @IlifilzaRusli Hey, the approach that i used is this.
There is a class ( say A ) that is build on top of the react-native-modal, every time some class ( say B ) needs a modal to be shown it calls class A. Class A puts the instance of class B into a queue. Now if another class ( say c ) needs a modal, it again calls A. A then checks if there is any other instance that is open. It sees that B is already open, it then closes the modal B deletes its instance and opens C and saves its instance. This cycle keeps on going.

Sorry for the delay, by this weekend i'll be making the entire code public.

**Comment by mmazzarolo** [comment]
Multiple modals might work using [the `portal` branch](https://github.com/react-native-community/react-native-modal/issues/145#issuecomment-421751346).  
Anyone willing to give it a try?

**Comment by odenlerma** [comment]
mine worked by using onModalHide then setTimeout on the setState of the next modal

**Comment by kalwaniyash** [comment]
@jojonarte @IlifilzaRusli
Check out this library [@kalwani/react-native-modal](https://www.npmjs.com/package/@kalwani/react-native-modal)
This will do the trick for you. It has been tested in production as well. Let me know if it works out for you guys or if you are facing any issues with it.

**Comment by hpcarlos** [comment]
@kalwaniyash Are you going to send this proposal to the princial branch?
it is working normally for me.
I have not tested on production yet, but I should test in a few days.

**Comment by scottmas** [comment]
@kalwaniyash This solution is gold 👍. I've been fighting with this for ages now. Where's your github repo though? Either you or someone else need to merge in this functionality.

**Comment by mmazzarolo** [comment]
Closing, you can check the "Multiple Modals" section in the README.md 👍

**Comment by plaa** [comment]
As another workaround, I created a `MutexModal` component (code below) that uses Redux state to only allow a single modal to be displayed at a time. The modals race to lock the mutex and the first one to acquire the lock (in practice the first one rendered) wins.

Usage: `<MutexModal name="myModalName" {...otherModalProps}>...</MutexModal>`

The name is used as the lock, and if not provided it works as a regular `Modal`. (This probably could be skipped, but it makes it easier to debug the Redux store.)

It would be great if this kind of functionality could be implemented in the `react-native-modal` library, because then multiple modals would _just work_ instead of leaving synchronization up to the dev. I'm not sure though what to use as a context-agnostic replacement for the Redux state.

<details><summary>MutexModal.tsx</summary>

```tsx
import React, { useEffect } from 'react';
import Modal, { ModalProps } from 'react-native-modal';
import { commonConstants } from '../styles/layout';
import { RootReducerState } from '../../store/types';
import {
  lockVisibleModal as lockVisibleModalAction,
  releaseVisibleModal as releaseVisibleModalAction,
} from '../../store/ui/UiActions';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

interface OwnProps extends ModalProps {
  name?: string;
}

interface StateProps {
  visibleModal: string | null;
}

interface DispatchProps {
  lockVisibleModal: (name: string) => void;
  releaseVisibleModal: (name: string) => void;
}

type Props = OwnProps & StateProps & DispatchProps;

/**
 * A wrapper for Modal which displays at most one modal at a time. If multiple
 * modals are displayed at a time, the latter modals never appear.
 * https://github.com/react-native-community/react-native-modal/issues/171
 *
 * If the 'name' prop is not passed, it works just as Modal.
 */
const MutexModal = ({
  isVisible,
  name,
  visibleModal,
  lockVisibleModal,
  releaseVisibleModal,
  ...props
}: Props) => {
  useEffect(() => {
    if (name) {
      if (isVisible && !visibleModal) {
        // Grab modal mutex when available
        lockVisibleModal(name);
      }
      if (!isVisible && visibleModal === name) {
        // Wait for modal hide animation before releasing mutex
        setTimeout(() => releaseVisibleModal(name), commonConstants.delayHideModal);
      }
    }
  }, [isVisible, visibleModal, name, lockVisibleModal, releaseVisibleModal]);

  const visible = name ? isVisible && visibleModal === name : isVisible;
  return <Modal isVisible={visible} {...props} />;
};

const mapStateToProps = (state: RootReducerState): StateProps => {
  const visibleModal = state.ui.visibleModal;
  return {
    visibleModal,
  };
};

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
  lockVisibleModal: (modal: string) => dispatch(lockVisibleModalAction(modal)),
  releaseVisibleModal: (modal: string) => dispatch(releaseVisibleModalAction(modal)),
});

export default connect(mapStateToProps, mapDispatchToProps)(MutexModal);
```

</details>

<details><summary>UiReducer.ts</summary>

```ts
    case ActionTypes.LOCK_VISIBLE_MODAL: {
      return {
        ...state,
        visibleModal: state.visibleModal ?? action.payload.modal,
      };
    }
    case ActionTypes.RELEASE_VISIBLE_MODAL: {
      return {
        ...state,
        visibleModal: state.visibleModal === action.payload.modal ? null : state.visibleModal,
      };
    }
```

</details>

**Comment by JPSantistebanQ** [comment]
I fixed this with conditional rendering:

{
				showMetodo ? <ModalMetodoPago
					showMetodo={showMetodo}
					hiddenModal={() => setShowMetodo(false)}
					onSelect={() => {
						setShowMetodo(false)
						setShowMetodoDetalle(true)
					}}
				/> : null
			}
			{
				showMetodoDetalle ? <ModalMetodoPagoDetalle
					showMetodo={showMetodoDetalle}
					hiddenModal={() => {
						console.warn('Hide modal')
						setShowMetodoDetalle(false)
						setShowMetodo(true)

					}}
					onSelect={() => {
						setShowMetodoDetalle(false)
					}}
				/> : null
			}

**Comment by bharatmodi777** [comment]
> @jojonarte @IlifilzaRusli
> Check out this library [@kalwani/react-native-modal](https://www.npmjs.com/package/@kalwani/react-native-modal)
> This will do the trick for you. It has been tested in production as well. Let me know if it works out for you guys or if you are facing any issues with it.

@kalwaniyash I have to use the solution you have provided, i'm new to React Native, can you please explain the usage of props? 

My use case is i have to handle session timeout at my app side so i'm setting timeout in the root file and showing the Modal on timeout. The issue i'm facing is, if some other screen is showing the Modal to the user and if user doesn't take any action on that modal and at the same time if my session is expired then in that case my session timeout Modal is not getting displayed. 

So i think your wrapper can help me in this case but i'm not able to use it properly.

**Comment by 452MJ** [comment]
Just try. It will work.
```jsx
<Modal_1>
   <Modal_2/>
</Modal_1>
```

**Comment by Saqib92** [comment]
Flickering happen if using react-native picker inside modal. 
https://github.com/n4kz/react-native-material-dropdown

**Comment by vishwa1937** [comment]
> Just try embed ModalB in ModalA. It will work.
> 
> ```js
> <ModalA>
>    <ModalB/>
> </ModalA>
> ```

This worked. Thanks, @452MJ .

**Comment by LESANF** [comment]
> Just try embed ModalB in ModalA. It will work.
> 
> ```js
> <ModalA>
>    <ModalB/>
> </ModalA>
> ```

thanks

**Comment by hotaryuzaki** [comment]
i have a similar issue, 1st modal i set to false then 2nd modal i set to true sequentially.
the issue is the 2nd modal not showing and the screen cannot response to any touch.
my solution is call the second modal after animation is done like this

```
setModal1(false);
InteractionManager.runAfterInteractions(async () => {
     setModal2(true);
});
```

**Comment by rabbitmouse** [comment]
> i have a similar issue, 1st modal i set to false then 2nd modal i set to true sequentially. the issue is the 2nd modal not showing and the screen cannot response to any touch. my solution is call the second modal after animation is done like this
> 
> ```
> setModal1(false);
> InteractionManager.runAfterInteractions(async () => {
>      setModal2(true);
> });
> ```

You can use setTimeout instead of runAfterInteractions, it works but is not a perfect solution

I also want to know what is the best way to switch between two different modals.

**Comment by hotaryuzaki** [comment]
> > i have a similar issue, 1st modal i set to false then 2nd modal i set to true sequentially. the issue is the 2nd modal not showing and the screen cannot response to any touch. my solution is call the second modal after animation is done like this
> > ```
> > setModal1(false);
> > InteractionManager.runAfterInteractions(async () => {
> >      setModal2(true);
> > });
> > ```
> 
> You can use setTimeout instead of runAfterInteractions, it works but is not a perfect solution
> 
> I also want to know what is the best way to switch between two different modals.

No, in most cases, the best way is to use InteractionManager.
Yes, you can change it with setTimeout, but you have to try a few times to find the perfect timing to make sure the issue won't occur anymore.
The problem is that, is hard to find the perfect timing because each phone will have a different time consumed to animate. For Android you will have a wide of range difference

**Comment by travis575757** [comment]
I needed modals to work anywhere in the app so the nested modal approach didn't work.  Here is a context based method that basically provides the vanilla modal experience but works on iOS just fine.

Define a modal context which manages the modal instances

```javascript
const StackableModalContext = createContext({
  mountModal: () => {},
});

/**
 * Provider for modals which can be rendered simultaneously (stacked)
 */
function StackableModalProvider({ children }) {
  const keysRef = useRef([]);
  const stateRef = useRef([]);
  const onShowListRef = useRef([]);
  const [state, setState] = useState([]);

  useEffect(() => {
    if (state.length > 0) {
      // invoke the onShow callback if it exists for this view
      onShowListRef.current.forEach(callback => {
        callback();
      });
      onShowListRef.current = [];
    }
  }, [state]);

  const mountModal = useCallback((modalChildren) => {
    // compute a random unique identifier
    const timestamp = Date.now();
    const randomNum = Math.random();
    const id = `${timestamp.toString(36)}-${randomNum.toString(36).substr(2, 9)}`;
    // push the modal data
    stateRef.current.push(modalChildren);
    keysRef.current.push(id);
    setState([...stateRef.current]);

    return [
      () => {
        const idx = keysRef.current.indexOf(id);
        keysRef.current.splice(idx, 1);
        stateRef.current.splice(idx, 1);
        setState([...stateRef.current]);
      },
      (updatedChildren) => {
        const idx = keysRef.current.indexOf(id);
        stateRef.current[idx] = updatedChildren;
        setState([...stateRef.current]);
      },
    ];
  }, []);

  return (
    <StackableModalContext.Provider value={{ mountModal }}>
      {children}
      {state.map((item, index) => (
        <React.Fragment key={index}>{item}</React.Fragment>
      ))}
    </StackableModalContext.Provider>
  );
}
```

Define a new base modal element which uses this context.  This is the component that acts like a normal react `Modal` component
```javascript
/**
 * Component to use as a parent for all modals
 */
function StackableModal({ children, visible }) {
  console.log(useContext)
  console.log(StackableModalContext)
  const { mountModal } = useContext(StackableModalContext);
  const updateRef = useRef(null);

  useEffect(() => {
    if (visible) {
      const handler = BackHandler.addEventListener('hardwareBackPress', () => true);
      return () => handler.remove();
    }
  }, [visible]);

  function getView(children) {
    return (
      <View
        pointerEvents={visible ? 'auto' : 'none'}
        style={{ position: 'absolute', width: '100%', height: '100%', opacity: visible ? 1 : 0 }}
      >
        {children}
      </View>
    );
  }

  useEffect(() => {
    const [unmount, update] = mountModal(getView(children));
    updateRef.current = update;
    return unmount;
  }, []);

  useEffect(() => {
    if (updateRef.current) {
      updateRef.current(getView(children));
    }
  }, [children]);

  return null;
}

```

Now you can basically use it like a normal react modal (obviously it doesn't have all the same props and stuff but that can be added as you need)

```javascript
function MyCustomModal(props) {
  return (
    <StackableModal visible={props.visible}>
      <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center', backgroundColor: 'rgba(0,0,0,0.5)' }}>
        <View style={{ backgroundColor: 'white', padding: 20, borderRadius: 10, width: '80%', alignItems: 'center' }}>
          <Text style={{ marginBottom: 20 }}>Hello, this is a modal!</Text>
          <Button title="Close Modal" onPress={() => {
            props.setVisible(false);
          }} />
        </View>
      </View>
    </StackableModal>
  );
}

export default function App() {
  const [visible1, setVisible1] = useState(false);
  const [visible2, setVisible2] = useState(false);
  const [visible3, setVisible3] = useState(false);
  return (
    <StackableModalProvider>
      <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
        <Text>Main Application View</Text>
        <MyCustomModal title="modal 1" visible={visible1} setVisible={setVisible1} />
        <MyCustomModal title="modal 2" visible={visible2} setVisible={setVisible2} />
        <MyCustomModal title="modal 3" visible={visible3} setVisible={setVisible3} />
        <Button title='show' onPress={() => {
          setVisible1(true)
          setVisible2(true)
          setVisible3(true)
        }} />
      </View>
    </StackableModalProvider>
  );
}
```

Demo Here: https://snack.expo.dev/@travis575757/993dec
